client/src/modules/Interviews/pages/feedback/feedbackTemplateHandler.ts

```diff
@@ -0,0 +1,231 @@
+ */
+export function getDefaultStep(feedback: Feedback) {
+  for (let i = 0; i < feedback.steps.length; i++) {
+    const { isCompleted, id, values } = feedback.steps[i];
+
+    if (!isCompleted || isInterviewRejected(id, values)) {
+      return i;
+    }
+  }
+
+  return feedback.isCompleted ? feedback.steps.length - 1 : 0;
+}
+
+/**
+ * checks whether the step contains rejection value
+ */
+export function isInterviewRejected(stepId: FeedbackStepId, stepValues: InterviewFeedbackValues = {}) {
+  return stepId === FeedbackStepId.Introduction && stepValues.interviewResult === 'missed';
+}
+
+/**
+ * Merges save feedback data with template
+ */
+function mergeFeedbackValuesToTemplate(feedback: Feedback, data: FeedbackData, interviewMaxScore: number): Feedback {
+  const { steps } = data;
+
+  const mergedFeedback = {
+    ...feedback,
+    steps: feedback.steps.map(step => {
+      const stepData = steps[step.id];
+      const result = {
+        ...step,
+        ...stepData,
+      };
+      const ratedSteps = [FeedbackStepId.Theory, FeedbackStepId.Practice];
+
+      if (ratedSteps.includes(result.id) && result.values) {
+        result.values.score = calculateStepScore(result, interviewMaxScore / ratedSteps.length);
+      }
+
+      return result;
+    }),
+  };
+
+  return applyDefaultFinalScore(mergedFeedback, interviewMaxScore);
+}
+
+function applyDefaultFinalScore(mergedFeedback: Feedback, interviewMaxScore: number) {
+  return {
+    ...mergedFeedback,
+    steps: mergedFeedback.steps.map(step => {
+      if (step.id !== FeedbackStepId.Decision) {
+        return step;
+      }
+      return {
+        ...step,
+        items: step.items.map(item => {
+          if (item.id === 'finalScore') {
+            return {
+              ...item,
+              defaultValue: calculateFinalScore(mergedFeedback.steps),
+              max: interviewMaxScore,
+            };
+          }
+          return item;
+        }),
+      };
+    }),
+  };
+}
+
+export function getUpdatedFeedback({
+  activeStepIndex,
+  feedback,
+  interviewMaxScore,
+  newValues,
+}: {
+  feedback: Feedback;
+  newValues: InterviewFeedbackValues;
+  activeStepIndex: number;
+  interviewMaxScore: number;
+}) {
+  const { steps } = feedback;
+  const isRejected = isInterviewRejected(steps[activeStepIndex].id, newValues);
+
+  const feedbackValues = {
+    steps: steps.reduce((stepMap, step, index) => {
+      if (index === activeStepIndex) {
+        stepMap[step.id] = {
+          isCompleted: true,
+          values: newValues,
+        };
+
+        return stepMap;
+      }
+
+      // if is rejected, all steps after the current one should be marked as not completed and values should be removed
+      stepMap[step.id] = {
+        values: isRejected ? undefined : step.values,
+        isCompleted: isRejected ? false : step.isCompleted,
+      };
+
+      return stepMap;
+    }, {} as Record<FeedbackStepId, InterviewFeedbackStepData>),
```

Comment:
Maybe it makes sense to extract this into a separate function?
```
function generateFeedbackValues(steps: FeedbackStep[], activeStepIndex: number, newValues: InterviewFeedbackValues, isRejected: boolean): Record<FeedbackStepId, InterviewFeedbackStepData> {
  return steps.reduce((stepMap, step, index) => {
    if (index === activeStepIndex) {
      stepMap[step.id] = {
        isCompleted: true,
        values: newValues,
      };
      return stepMap;
    }

    stepMap[step.id] = {
      values: isRejected ? undefined : step.values,
      isCompleted: isRejected ? false : step.isCompleted,
    };
    return stepMap;
  }, {} as Record<FeedbackStepId, InterviewFeedbackStepData>);
}
```

Reply:
sure